File: app/src/main/java/com/migue/zeus/expensesnotes/data/models/Expense.java
package com.migue.zeus.expensesnotes.data.models;


import android.arch.persistence.room.Entity;
import android.arch.persistence.room.ForeignKey;
import android.arch.persistence.room.Index;
import android.arch.persistence.room.PrimaryKey;
import android.content.Context;

import com.migue.zeus.expensesnotes.R;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;


@Entity(
        tableName = "Expenses",
        foreignKeys = {
                @ForeignKey(entity = Icon.class, parentColumns = "Id", childColumns = "IconId"),
                @ForeignKey(entity = ExpenseCategory.class, parentColumns = "Id", childColumns = "ExpenseCategoryId")
        },
        indices = {
                @Index("IconId"),
                @Index("ExpenseCategoryId")
        })
public class Expense {
    @PrimaryKey(autoGenerate = true)
    private int Id;
    private String Name;
    private double Value;
    private Date CreationDate;
    private Date ModificationDate;
    private int ExpenseCategoryId;
    private int IconId;

    public Expense(String name, double value, Date creationDate, int expenseCategoryId) {
        Name = name;
        Value = value;
        CreationDate = creationDate;
        ExpenseCategoryId = expenseCategoryId;
        IconId = 1;
    }

    public Expense() {
    }

    public int getIconId() {
        return IconId;
    }

    public void setIconId(int iconId) {
        IconId = iconId;
    }

    public int getExpenseCategoryId() {
        return ExpenseCategoryId;
    }

    public void setExpenseCategoryId(int ExpenseCategoryId) {
        this.ExpenseCategoryId = ExpenseCategoryId;
    }

    public int getId() {
        return Id;
    }

    public void setId(int id) {
        Id = id;
    }

    public String getName() {
        return Name;
    }

    public void setName(String name) {
        Name = name;
    }

    public double getValue() {
        return Value;
    }

    public void setValue(double value) {
        Value = value;
    }

    public Date getCreationDate() {
        return CreationDate;
    }

    public void setCreationDate(Date creationDate) {
        CreationDate = creationDate;
    }

    public Date getModificationDate() {
        return ModificationDate;
    }

    public void setModificationDate(Date modificationDate) {
        ModificationDate = modificationDate;
    }

    public static Date parseDate(String date) {
        try {
            return new SimpleDateFormat("yyyy-MM-dd", Locale.US).parse(date);
        } catch (ParseException e) {
            return null;
        }
    }
    public static Expense[] populateData() {

        return new Expense[]{
                new Expense("Expense a", 367.0, parseDate("2018-12-01"), 1),
                new Expense("Expense b", 367.0, parseDate("2018-12-01"), 1),
                new Expense("Expense c", 367.0, parseDate("2018-12-05"), 1),
                new Expense("Expense d", 367.0, parseDate("2018-12-07"), 1),
                new Expense("Expense e", 367.0, parseDate("2018-12-07"), 1),
                new Expense("Expense f", 367.0, parseDate("2018-12-10"), 1),
                new Expense("Expense g", 367.0, parseDate("2018-12-10"), 1),
                new Expense("Expense h", 367.0, parseDate("2018-12-10"), 1),
                new Expense("Expense i", 367.0, parseDate("2018-12-10"), 1),
                new Expense("Expense j", 367.0, parseDate("2018-12-10"), 1),
                new Expense("Expense k", 367.0, parseDate("2018-12-12"), 1),
                new Expense("Expense l", 367.0, parseDate("2018-12-12"), 1),
                new Expense("Expense m", 367.0, parseDate("2018-12-15"), 1),
                new Expense("Expense n", 367.0, parseDate("2018-12-15"), 1),
                new Expense("Expense o", 367.0, parseDate("2018-12-16"), 1),
                new Expense("Expense p", 367.0, parseDate("2018-12-16"), 1),
                new Expense("Expense q", 367.0, parseDate("2018-12-17"), 1),
                new Expense("Expense r", 367.0, parseDate("2018-12-18"), 1),
                new Expense("Expense s", 367.0, parseDate("2018-12-18"), 1),
                new Expense("Expense t", 367.0, parseDate("2018-12-18"), 1),
                new Expense("Expense u", 367.0, parseDate("2018-12-18"), 1),
                new Expense("Expense v", 367.0, parseDate("2018-12-18"), 1),
                new Expense("Expense w", 367.0, parseDate("2018-12-19"), 1),
                new Expense("Expense x", 367.0, parseDate("2018-12-22"), 1),
                new Expense("Expense y", 367.0, parseDate("2018-12-26"), 1),
                new Expense("Expense z", 367.0, parseDate("2018-12-29"), 1),
        };
    }
}

File: app/src/main/java/com/migue/zeus/expensesnotes/data/models/Expense.kt
package com.migue.zeus.expensesnotes.data.models


import android.arch.persistence.room.*
import android.os.Parcel
import android.os.Parcelable
import com.migue.zeus.expensesnotes.infrastructure.utils.Converters.parseDate

import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale


@Entity(tableName = "Expenses",
        foreignKeys = [ForeignKey(entity = ExpenseCategory::class, parentColumns = ["Id"], childColumns = ["ExpenseCategoryId"])],
        indices = [Index("ExpenseCategoryId")])
class Expense : BaseEntity {
    override fun getReadableName(): String? {
        return name
    }
    @ColumnInfo(name = "Name")
    var name: String? = null
    @ColumnInfo(name = "ExpenseCategoryId")
    var expenseCategoryId: Long? = null
    @ColumnInfo(name = "Date")
    var date: Date? = Date()

    constructor(parcel: Parcel) : super(parcel) {
        name = parcel.readString()
        expenseCategoryId = parcel.readValue(Long::class.java.classLoader) as? Long
        date = parcel.readValue(Date::class.java.classLoader) as? Date
    }

    constructor()

    @Ignore
    constructor(name: String, date: Date?, expenseCategoryId: Long) {
        this.expenseCategoryId = expenseCategoryId
        this.name = name
        this.date = date
    }


    override fun writeToParcel(parcel: Parcel, flags: Int) {
        super.writeToParcel(parcel, flags)
        parcel.writeString(name)
        parcel.writeValue(expenseCategoryId)
        parcel.writeValue(date)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<Expense> {
        override fun createFromParcel(parcel: Parcel): Expense {
            return Expense(parcel)
        }

        override fun newArray(size: Int): Array<Expense?> {
            return arrayOfNulls(size)
        }

        fun populateData(): Array<Expense> {

            return arrayOf(
                    Expense("Expense a", parseDate("2018-12-01"), 1),
                    Expense("Expense b", parseDate("2018-12-01"), 2),
                    Expense("Expense c", parseDate("2018-12-05"), 3),
                    Expense("Expense d", parseDate("2018-12-07"), 4),
                    Expense("Expense e", parseDate("2018-12-07"), 1),
                    Expense("Expense f", parseDate("2018-12-10"), 1),
                    Expense("Expense g", parseDate("2018-12-10"), 1),
                    Expense("Expense h", parseDate("2018-12-10"), 1),
                    Expense("Expense i", parseDate("2018-12-10"), 1),
                    Expense("Expense j", parseDate("2018-12-10"), 1),
                    Expense("Expense k", parseDate("2018-12-12"), 1),
                    Expense("Expense l", parseDate("2018-12-12"), 1),
                    Expense("Expense m", parseDate("2018-12-15"), 1),
                    Expense("Expense n", parseDate("2018-12-15"), 1),
                    Expense("Expense o", parseDate("2018-12-16"), 1),
                    Expense("Expense p", parseDate("2018-12-16"), 1),
                    Expense("Expense q", parseDate("2018-12-17"), 1),
                    Expense("Expense r", parseDate("2018-12-18"), 1),
                    Expense("Expense s", parseDate("2018-12-18"), 1),
                    Expense("Expense t", parseDate("2018-12-18"), 1),
                    Expense("Expense u", parseDate("2018-12-18"), 1),
                    Expense("Expense v", parseDate("2018-12-18"), 1),
                    Expense("Expense w", parseDate("2018-12-19"), 1),
                    Expense("Expense x", parseDate("2018-12-22"), 1),
                    Expense("Expense y", parseDate("2018-12-26"), 1),
                    Expense("Expense z", parseDate("2018-12-29"), 1)
            )
        }
    }

}